### ISSUE TITLE
Category imbalance in relation to AP values

### ISSUE THREAD

**Issue by xcaizewu** [fizyr/keras-retinanet#1074]
I used retinanet to train my own data set, using the default Settings (only changed the number of iterations and learning rate and other parameters).

My training set categories were unbalanced. From the first epoch to the fifth epoch, the AP value of a small number of categories was very low, but I trained the same data set on faster RCNN, and this situation did not occur. Can someone help me figure out why this is **happening?（The first picture is the AP value obtained after the fifth epoch, and the second picture is the AP value obtained after faster RCNN ran out. The faster RCNN only has the number and no category name, but they can be matched）**

![(46{AJTE ~ZRUV4R(C~89NT](https://user-images.githubusercontent.com/48117810/61343713-5877bb00-a881-11e9-8df4-47cede30ce75.png)

![B1AGEB% I% 2P)R II`9~U](https://user-images.githubusercontent.com/48117810/61343785-9a086600-a881-11e9-9fa0-2bfcaab7daf7.png)

**Comment by hgaiser** [comment]
Retinanet might need longer to train than FasterRCNN. Did you train for longer, until the mAP score stabilized? Also, what was your starting weights; imagenet or coco?

**Comment by xcaizewu** [comment]
@hgaiser 

Thank you for answering all my questions. I used imagenet, and the training speed was too slow. I took an epoch for 80,000 times, which took about one day to train, so I only trained the fifth epoch, and did not train any more. Perhaps this is why my mAP value is worse than that of the faster- RCNN. But since the first round, there has been little improvement in categories with low AP scores, such as watch fruits and glassbos. In other categories, AP values are more or less raised.

**Comment by hgaiser** [comment]
Could you try with fewer steps, say 5k?

**Comment by hgaiser** [comment]
Any update on this ?

**Comment by xcaizewu** [comment]
The AP value of 5k is about the same. I am still in training, but the AP value of watch is not improved at all

![$1LDQNXFE_4GU$ _IQIFJYU](https://user-images.githubusercontent.com/48117810/62118280-fd69ad80-b2ef-11e9-853a-0efb75730c0f.jpg)

**Comment by hgaiser** [comment]
Also, what is your tensorflow version? Could you try updating to 1.14?

**Comment by xcaizewu** [comment]
20/5000  
I've been using tesorflow 1.8 and will try to upgrade to 1.14

**Comment by xcaizewu** [comment]
@hgaiser 

I've been using tesorflow 1.8 and will try to upgrade to 1.14

**Comment by hgaiser** [comment]
Did updating work?

**Comment by xcaizewu** [comment]
@hgaiser 

Sorry, I have given up using retinanet. I have to upgrade cuda to 10 using tensorflow1.14, but there are other projects running on my server that cannot upgrade cuda, so I plan to continue using Faster-rcnn.

**Comment by aojue1109** [comment]
many label AP is 0.000，epoch=50 tensorflow=1.14.0

20 instances of class 抵押担保函-1 with average precision: 0.0500
20 instances of class 抵押担保函-2 with average precision: 0.9413
20 instances of class 抵押担保函-3 with average precision: 0.1000
0 instances of class 抵押担保函-stamp with average precision: 0.0000
12 instances of class 收款确认函-1-1 with average precision: 0.8963
12 instances of class 收款确认函-1-2 with average precision: 0.8830
12 instances of class 收款确认函-1-3 with average precision: 0.9167
0 instances of class 收款确认函-1-4 with average precision: 0.0000
5 instances of class 收款确认函-2-1 with average precision: 0.0000
5 instances of class 收款确认函-2-2 with average precision: 0.9667
5 instances of class 收款确认函-2-3 with average precision: 1.0000
1 instances of class 收款确认函-3-1 with average precision: 1.0000
1 instances of class 收款确认函-3-3 with average precision: 1.0000
20 instances of class 融资租赁合同-1 with average precision: 0.0167
20 instances of class 融资租赁合同-2 with average precision: 1.0000
20 instances of class 融资租赁合同-3 with average precision: 0.9360
20 instances of class 融资租赁合同-4 with average precision: 0.0000
20 instances of class 融资租赁合同-5 with average precision: 0.0000
20 instances of class 融资租赁合同-6 with average precision: 0.0000
20 instances of class 融资租赁合同-7 with average precision: 0.0000
20 instances of class 融资租赁合同-8 with average precision: 0.9406
20 instances of class 融资租赁合同-9 with average precision: 0.9452
28 instances of class 融资租赁合同-stamp with average precision: 0.9509

**Comment by hgaiser** [comment]
Closing this as there is no issue left.

**Comment by mary-0830** [comment]
> I used retinanet to train my own data set, using the default Settings (only changed the number of iterations and learning rate and other parameters).
> 
> My training set categories were unbalanced. From the first epoch to the fifth epoch, the AP value of a small number of categories was very low, but I trained the same data set on faster RCNN, and this situation did not occur. Can someone help me figure out why this is **happening?（The first picture is the AP value obtained after the fifth epoch, and the second picture is the AP value obtained after faster RCNN ran out. The faster RCNN only has the number and no category name, but they can be matched）**
> 
> ![(46{AJTE ZRUV4R(C89NT](https://user-images.githubusercontent.com/48117810/61343713-5877bb00-a881-11e9-8df4-47cede30ce75.png)
> 
> ![B1AGEB% I% 2P)R II`9~U](https://user-images.githubusercontent.com/48117810/61343785-9a086600-a881-11e9-9fa0-2bfcaab7daf7.png)

hello,May I ask what command do you use to get the ap value of each category? thank you for your reply.
